### PR TITLE
Adding web app and web app call web api CIAM non CUD tests

### PR DESCRIPTION
**Pull request by trwalke**
# Adding web app and web app call web api CIAM non CUD tests

- [ ] You've read the [Contributor Guide](https://github.com/AzureAD/microsoft-identity-web/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/AzureAD/microsoft-identity-web/blob/master/CODE_OF_CONDUCT.md).
- [ ] You've included unit or integration tests for your change, where applicable.
- [ ] You've included inline docs for your change, where applicable.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

Summary of the changes (Less than 80 chars)

## Description
Work in progress.
https://github.com/AzureAD/microsoft-identity-web/issues/2698

Adding web app and web app call web api CIAM non CUD tests
Web app calls web api
web app calls graph
Deamon app calls graph
pending lab api integration

{Detail}

Fixes #{bug number} (in this specific format)

### CODE REVIEW

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json [hunk 1]
```diff
@@ -17,15 +17,15 @@
...
-            {
-                "SourceType": "KeyVault",
-                "KeyVaultUrl": "https://webappsapistests.vault.azure.net",
-                "KeyVaultCertificateName": "Self-Signed-5-5-22"
-            }
-            // {
-            //    "SourceType": "ClientSecret",
-            //    "ClientSecret": ""
-            // }
+            //{
```
Comment (trwalke) [inline tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:20]: temporary change for testing. will create Cert in lab for CIAM app

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs [hunk 2]
```diff
@@ -172,6 +176,120 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPasswordCreds
                 playwright.Dispose();
             }
         }
+
+        [Fact]
+        [SupportedOSPlatform("windows")]
+        public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPasswordCreds_TodoAppFunctionsCorrectlyWithCiam()
```
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:182]: What is the difference between this test and the prev one? Does this call the web api and the prev one just "signs in"

**Comment by jennyf19** [comment]
Please make sure the branch is updated and would like @JoshLozensky to review before this is merged. They will be OOF until Monday, just FYI. Thanks.

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json (hunk 1 above)
Comment (JoshLozensky) [reply tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:20]: Has this run successfully with a hard-coded cert? I'm not able to run it myself without a secret.

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs [hunk 3]
```diff
@@ -103,5 +103,86 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPassword()
...
+    [Fact]
+    [SupportedOSPlatform("windows")]
+    public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailWithCiamPassword()
+    {
+        // Arrange
+        const string _clientSecret = "...";
+        Process? process = null;
+        const string TraceFileName = TraceFileClassName + "_ValidEmailPassword";
+        using IPlaywright playwright = await Playwright.CreateAsync();
+        IBrowser browser = await playwright.Chromium.LaunchAsync(new() { Headless = false });
```
Comment (JoshLozensky) [inline tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:116]: Probably best to default headless to true so tests run faster and there aren't browsers popping up when people run Id Web's tests.

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs [hunk 4]
```diff
@@ -103,5 +103,86 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPassword()
...
     }
+
+    [Fact]
+    [SupportedOSPlatform("windows")]
+    public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailWithCiamPassword()
+    {
+        // Arrange
+        const string _clientSecret = "...";
+        Process? process = null;
+        const string TraceFileName = TraceFileClassName + "_ValidEmailPassword";
```
Comment (JoshLozensky) [inline tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:114]: This needs to be changed or it will overwrite the other test's trace files. Each test should have a unique name.

**Comment by JoshLozensky** [comment]
If we are mostly doing the same tasks with the same programs, should we instead make these theories and split out the unique fields into theory data?

**Review by JoshLozensky — CHANGES_REQUESTED** [review]
Additionally, since these tests can be transiently unstable in the build pipeline while appearing fine locally, please test them multiple times if not adding a skip.

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 2 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:182]: Both of the tests call the down stream api, the previous test simply adds additional clicks in the app for the todo list feature which is not present in this app.

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json (hunk 1 above)
Comment (trwalke) [reply tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:20]: Not with cert but it does with a secret, will be updating with new app/credentials shortly

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 2 above)
Comment (bgavrilMS) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:182]: Is it worth having both? Is one a superset of the other?

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs (hunk 3 above)
Comment (bgavrilMS) [reply tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:116]: For sure, this probably just a forgotten setting.

File: tests/E2E Tests/TokenAcquirerTests/TokenAcquirer.cs [hunk 5]
```diff
@@ -92,6 +99,24 @@ public async Task AcquireToken_WithMicrosoftIdentityApplicationOptions_ClientCre
...
+        [IgnoreOnAzureDevopsFact]
+        //[Fact]
+        public async Task AcquireToken_WithMicrosoftIdentityApplicationOptions_ClientCredentialsCiamAsync()
+        {
+            TokenAcquirerFactory tokenAcquirerFactory = TokenAcquirerFactory.GetDefaultInstance();
+            IServiceCollection services = tokenAcquirerFactory.Services;
+
+            services.Configure<MicrosoftIdentityApplicationOptions>(s_optionName, option =>
+            {
+                option.Instance = "https://MSIDLABCIAM6.ciamlogin.com";
```
Comment (bgavrilMS) [inline tests/E2E Tests/TokenAcquirerTests/TokenAcquirer.cs:111]: With CIAM we would not be supporting the `Instance` and `TenantId` configuration. We expect ppl to configure `Authority` instead.

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs [hunk 6]
```diff
@@ -103,5 +103,86 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPassword()
...
+        IBrowser browser = await playwright.Chromium.LaunchAsync(new() { Headless = false });
+        IBrowserContext context = await browser.NewContextAsync(new BrowserNewContextOptions { IgnoreHTTPSErrors = true });
+        await context.Tracing.StartAsync(new() { Screenshots = true, Snapshots = true, Sources = true });
+
+        try
+        {
+            var clientEnvVars = new Dictionary<string, string>
+            {
+                {"AzureAd__ClientId", "16ee75bf-ebbe-42a8-9474-caa0fe856bc7"},//"Authority": "https://nativeauthasampleapp.ciamlogin.com"
+                {"AzureAd__Instance", "https://nativeauthasampleapp.ciamlogin.com"},
```
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:125]: Same as above, expecting the Authority to be set (for both CUD and non-CUD). Instance + TenantID not expected to work.

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json [hunk 7]
```diff
@@ -17,15 +17,15 @@
             //    "SourceType": "SignedAssertionFromManagedIdentity",
```
Comment (bgavrilMS) [inline tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:17]: @trwalke - Did you actually get the CIAM web app to get a token and call Graph? I am actually unsure if that scenario works today (it used to not work).

File: tests/DevApps/ciam/myWebApi/myWebApi.csproj [hunk 8]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFrameworks>net7.0</TargetFrameworks>
+    <TargetFrameworks>net8.0</TargetFrameworks>
```
Comment (bgavrilMS) [inline tests/DevApps/ciam/myWebApi/myWebApi.csproj:4]: unrelated?

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs [hunk 9]
```diff
@@ -25,20 +25,24 @@ public class WebAppCallsApiCallsGraphLocally : IClassFixture<InstallPlaywrightBr
...
         private readonly string _testAssemblyLocation = typeof(WebAppCallsApiCallsGraphLocally).Assembly.Location;
+        private readonly string _clientSecret = "...";
         private readonly ITestOutputHelper _output;
 
         public WebAppCallsApiCallsGraphLocally(ITestOutputHelper output)
         {
             _output = output;
         }
 
-        [Fact(Skip = "https://github.com/AzureAD/microsoft-identity-web/issues/2716")]
```
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:41]: Was this fixed ?

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 2 above)
Comment (bgavrilMS) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:182]: I am not sure this correct. I think the names are confusing. 

TestingWebAppLocally - tests just web app scenario (AcquireTokenByAuthcode / AcquireTokenSilent)
TestingWebAppCallsApiCallGraph - tests web app and web api scenario (so OBO too).

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs (hunk 3 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs:116]: Yes, this is for testing. will be changed back

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 2 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:182]: I agree, name will be updated. This test involves a web app calling a downstream web api (locally)

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json (hunk 7 above)
Comment (trwalke) [reply tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:17]: Yes, it worked when testing locally
image

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 9 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:41]: No, I will remove this change

File: tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json (hunk 7 above)
Comment (trwalke) [reply tests/DevApps/WebAppCallsMicrosoftGraph/appsettings.json:17]: however, see https://github.com/AzureAD/microsoft-identity-web/issues/2732

File: tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs (hunk 4 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/TestingWebAppLocally.cs]: Makes sense. Updated.

File: tests/DevApps/ciam/myWebApi/myWebApi.csproj (hunk 8 above)
Comment (trwalke) [reply tests/DevApps/ciam/myWebApi/myWebApi.csproj:4]: When the file path for the test apps is calculated, it expects the file to be in the .net 8 target folder. The CIAM app was using net 7 however so I figured the path of least resistance would be to update the app to .NET 8. App appears to be functioning the same.

Comment (bgavrilMS) [reply tests/DevApps/ciam/myWebApi/myWebApi.csproj:4]: Kind of brittle to have the tfm in the output dir...

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs [hunk 10]
```diff
@@ -172,6 +157,125 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPasswordCreds
...
+                _output.WriteLine("Web app sign-in flow successful using Sign in button after sign out.");
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail($"the UI automation failed: {ex} output: {ex.Message}.");
+            }
+            finally
+            {
+                // Add the following to make sure all processes and their children are stopped.
+                Queue<Process> processes = new();
```
Comment (bgavrilMS) [inline tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:236]: Can't this code be consolidated with other tests?

**Review by bgavrilMS — APPROVED** [review]

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs (hunk 10 above)
Comment (trwalke) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:236]: I was thinking about it but there are quite a few variables in this code block that would need to be passed into the consolidated method that it would end up being more complicated than its worth. I will take another look at it though.

File: tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs [hunk 11]
```diff
@@ -172,6 +157,125 @@ public async Task ChallengeUser_MicrosoftIdFlow_LocalApp_ValidEmailPasswordCreds
...
+                _output.WriteLine("Starting web app sign-in flow using sign in button after sign out.");
+                await page.GetByRole(AriaRole.Link, new() { Name = "Sign in" }).ClickAsync();
+                await UiTestHelpers.FirstLogin_MicrosoftIdFlow_ValidEmailPassword(page, email, LabUserHelper.FetchUserPassword("msidlabciam6"), _output);
+                await Assertions.Expect(page.GetByText("Welcome")).ToBeVisibleAsync(_assertVisibleOptions);
+                await Assertions.Expect(page.GetByText(email)).ToBeVisibleAsync(_assertVisibleOptions);
+                _output.WriteLine("Web app sign-in flow successful using Sign in button after sign out.");
+            }
+            catch (Exception ex)
+            {
+                Assert.Fail($"the UI automation failed: {ex} output: {ex.Message}.");
```
[the remark this answers is not in the document]
Comment (JoshLozensky) [reply tests/E2E Tests/WebAppUiTests/WebAppCallsApiCallsGraphLocally.cs:231]: The point of the try, catch, finally is to make sure we shut down the processes we started in the finally. We still need to throw the exception from the catch or else we just swallow it right?

**Review by JoshLozensky — APPROVED** [review]

**Review by JoshLozensky — CHANGES_REQUESTED** [review]
Can you please run this through our ADO build pipeline to ensure it doesn't increase UI test flakiness? They currently fail at about a 20 to 25% rate. They only fail in the build pipeline, not locally.

File: tests/DevApps/ciam/myWebApp/myWebApp.csproj [hunk 12]
```diff
@@ -5,6 +5,7 @@
     <Nullable>enable</Nullable>
     <ImplicitUsings>enable</ImplicitUsings>
     <UserSecretsId>aspnet-myWebApp-23b8728b-775f-4796-8b63-e85cbaccc27c</UserSecretsId>
+    <UseWIP>true</UseWIP>
```
Comment (bgavrilMS) [inline tests/DevApps/ciam/myWebApp/myWebApp.csproj:8]: what does this do?
